### PR TITLE
`Development`: Separate checkout tasks from build script for local continuous integration

### PR DESCRIPTION
**Pull request by laurenzfb**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
This PR separates checkout tasks from the build script that is executed within a build job container for the local CI system. This facilitates a cleaner script and is the basis to later facilitate custom created build scripts.

### Description
Cleaned up createBuildScript() and removed redundant copy steps to populate the container with repositories.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- TS 3 or 6

1. As Instructor, create a programming exercise.
2. Verify build results for template and solution
3. As student, create submission for programming exercise
4. Verify build results

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

**Review by milljoniaer — APPROVED** [review]
tested on ts3, works as expected 👍

**Review by vinceclifford — APPROVED** [review]
Tested on ts3. Works as expected when the default programming language is equal to the one used in the programming exercise. Else the builds fail. A new issue will be created for this bug.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIContainerService.java [hunk 1]
```diff
@@ -207,22 +206,72 @@ public void stopContainer(String containerName) {
...
+            convertDosFilesToUnix(newDirectoryName, containerId);
         }
-        copyToContainer(buildScriptPath.toString(), buildJobContainerId);
-        renameDirectoryOrFile(buildJobContainerId, buildScriptPath.getFileName().toString(), "script.sh");
     }
 
     private void renameDirectoryOrFile(String containerId, String oldName, String newName) {
         ExecCreateCmdResponse renameDirectoryCmdResponse = dockerClient.execCreateCmd(containerId).withCmd("mv", oldName, newName).exec();
-        dockerClient.execStartCmd(renameDirectoryCmdResponse.getId()).exec(new ResultCallback.Adapter<>());
+        final CountDownLatch latch = new CountDownLatch(1);
```
Comment (reschandreas) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIContainerService.java:239]: the latch definition is duplicated, maybe you could generalize it to duplicate the code a bit.

**Review by reschandreas — DISMISSED** [review]
code lgtm, only one comment

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIContainerService.java (hunk 1 above)
Comment (laurenzfb) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIContainerService.java]: Moved the docker command execution to a method.

**Review by reschandreas — APPROVED** [review]
reapprove

**Review by vinceclifford — APPROVED** [review]
reapprove

**Review by milljoniaer — APPROVED** [review]
Re-approve

**Review by florian-glombik — APPROVED** [review]
The code looks good to me, much easier to understand than the previous solution!

**Pull request merged**